FWCore/Framework/src/ESProductResolver.cc

```diff
@@ -74,15 +77,20 @@ namespace edm {
         nonTransientAccessRequested_.store(true, std::memory_order_release);
       }
 
-      if UNLIKELY (!cacheIsValid()) {
-        cache_ = getAfterPrefetchImpl();
-        cacheIsValid_.store(true, std::memory_order_release);
+      auto cache = cache_.load();
+      if UNLIKELY (cache == kInvalid) {
+        // This safe even if multiple threads get in here simultaneously
+        // because cache_ is atomic and getAfterPrefetchImpl will return
+        // the same pointer on all threads for the same IOV.
+        // This is fast because the vast majority of the time only 1 thread per IOV
+        // will get in here so most of the time only 1 atomic operation.
+        cache_.store(getAfterPrefetchImpl());
+        cache = cache_.load();
```

Comment:
```suggestion
      cache = cache_ = getAfterPrefetchImpl();
```
this may allow the compiler to generate more optimized code.

Reply:
Done. Thanks, that is better.